llrt_core/src/modules/encoding/text_encoder.rs

```diff
@@ -16,43 +17,56 @@ impl TextEncoder {
     }
 
     #[qjs(get)]
-    fn encoding(&self) -> String {
-        "utf-8".to_string()
+    fn encoding(&self) -> &str {
+        "utf-8"
     }
 
-    pub fn encode<'js>(&self, ctx: Ctx<'js>, string: String) -> Result<Value<'js>> {
-        TypedArray::new(ctx, string.as_bytes()).map(|m| m.into_value())
+    pub fn encode<'js>(&self, ctx: Ctx<'js>, string: Opt<Value<'js>>) -> Result<Value<'js>> {
+        if let Some(string) = string.0 {
+            if let Some(string) = string.as_string() {
+                if let Ok(string) = string.to_string() {
+                    return TypedArray::new(ctx, string.as_bytes()).map(|m| m.into_value());
+                }
+            }
+        }
+
+        TypedArray::new(ctx.clone(), []).map(|m: TypedArray<'_, u8>| m.into_value())
     }
 
     pub fn encode_into<'js>(
         &self,
         ctx: Ctx<'js>,
-        string: String,
-        obj: Object<'js>,
+        src: String,
+        dst: Value<'js>,
     ) -> Result<Object<'js>> {
-        let mut read = 0;
-        let mut written = 0;
-
-        if let Some((array_buffer, source_length, source_offset)) = obj_to_array_buffer(&obj)? {
+        if let Ok(typed_array) = TypedArray::<u8>::from_value(dst) {
+            let dst_length = typed_array.len();
+            let dst_offset: usize = typed_array.get("byteOffset")?;
+            let array_buffer = typed_array.arraybuffer()?;
             let raw = array_buffer
                 .as_raw()
                 .ok_or("ArrayBuffer is detached")
                 .or_throw(&ctx)?;
 
-            let bytes = unsafe {
-                std::slice::from_raw_parts_mut(raw.ptr.as_ptr().add(source_offset), source_length)
+            let dst = unsafe {
+                std::slice::from_raw_parts_mut(raw.ptr.as_ptr().add(dst_offset), dst_length)
             };
 
             let mut enc = encoding_rs::UTF_8.new_encoder();
-            (_, _, written, _) = enc.encode_from_utf8(string.as_str(), bytes, false);
-            read = string[..written]
+            let (_, _, written, _) = enc.encode_from_utf8(&src, dst, false);
+            let read = src[..written]
                 .chars()
                 .fold(0, |acc, ch| acc + ch.len_utf16());
-        }
```

Comment:
Can we avoid iterating over each char to get the count?

Reply:
So...I spent quite a lot of time playing with a lot of different approaches here, from using `qjs` primitives to get the string bytes, recast as `&[u16]`, `encode_from_utf16()` to get the `read` and `written` correctly, and to handle all of the test cases around unicode and surrogates (and unpaired surrogates), to attempting to encode UTF-8 with fallback to UTF-16, but I couldn't get anything to fully work. According to [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder/encodeInto#read), `read` is the number of UTF-16 code points, even though we're encoding into UTF-8. There are improvements to be made here, but I would need to thoroughly read the spec and understand how those edge cases are supposed to be handled.

Reply:
Oh, I forgot to mention, I think the issue with surrogates and such is that `encoding_rs` only accepts a `&str` for `encode_from_utf8` which requires turning a byte array into a string slice, and Rust requires a string slice to be valid UTF-8. To handle those edge cases, I believe `encoding_rs` should accept a byte array and check for surrogates and whatever other edge cases.

Reply:
> `read` is the number of UTF-16 code points, even though we're encoding into UTF-8. 

When I first implemented this specification, it was the hardest thing for me to understand. No matter how much research I did, I could not find a use case for it, so I still don't understand it. :(

We may have to go back to the specification development process to understand it.

Apart from that, your refactor content confirms that I was not wrong to bring in `encoding_rs`. Thank you very much. :)

Reply:
@nabetti1720 Thanks for reviewing! Yeah, I think you did a wonderful job. We have our own basic implementation of `TextEncoder/Decoder` and we also used `encoding_rs`. I'm looking forward to switching ours out for this one once we get this PR merged so we don't have to maintain it ourselves. Would be cool to implement the streaming features at some point too.